### PR TITLE
Deploy to tenderly

### PR DESCRIPTION
**Pull request by anxolin**
The main motivation of this PR is to deploy to our tenderly account the WatchTower.

However, during the deployment, I had to do some minor changes due to some issues I observed. Sorry for a MISC PR, but just to make it easier I documented in the description the changes and also did some code annotations giving more context (see review mode).

## How to deploy this?
Since I didn't know much about this service, I created some documentation that I hope is handy for ONCALL and mantainers of this project.

It shows how to deploy, gives context of what it does. It Explains and gives pointers for all the related concepts you need to know to understand this project, so anyone can get up to speed.

Additionally this document will point out about the evolution (replacement by a distributed version)

https://www.notion.so/cownation/WatchTower-legacy-Tenderly-Composable-CoW-Indexer-8e62049feb6e440d9b097f83e054ac75?pvs=4

## Local test for Tenderly Web3 Actions
To debug some issues I had to run Tenderly Actions locally to shorten the dev cycle.

- I realized there was no instructions in the README, so I added a section
- The environment vars were not being loaded from `.env` file, so I added dotenv lib. I think it was working for the CLI but notfor the github actions. 

## Allow basic Auth in the EthersJS provider
Some RPC nodes require basic auth, so i had to add it.

## Allow basic Auth in the EthersJS provider
Now we support to add optionally a user and password for each 

```ini
NODE_USER_100=your-user
NODE_PASSWORD_100=your-password
```

## Reduce logs
I removed a bunch of logs, so we would write in the log only when there's relevant changes.
For example, we don't write any more "processing block XXX", but we log if we detect a new order, post to the API, etc.

Actually the logs were breaking the indexing (circular JSON dependency), see:
Screenshot at Aug 07 17-19-06

Also, it was making Tenderly to remove all logs and show you nothing:
Screenshot at Aug 07 19-04-40

# Improve log messages
The issue above was a bit hard to debug cause i was not sure which log was causing this. This PR in general adds some more context on where things fail, and give you some nice outputs of requests, API responses, etc:

Screenshot at Aug 08 09-36-52

# Not included
I still want to reiterate in other PRs to:
- Receive a notification when there's an error
- Make the execution fail if there was any errors (currently shows you all GREEN even when there was issues)
- Handle the duplicate order exception (so its not an error, and we just log that someone else posted the order)
- [Good to have] Maybe Handle the Settlement Allowance check, because right now we are posting orders that don't have the balance and allowance, so we rely on the backend to tell us. This will save them some resources....

### CODE REVIEW

File: actions/watch.ts [hunk 1]
```diff
@@ -83,101 +97,123 @@ export const checkForAndPlaceOrder: ActionFn = async (
...
-            ...orderToSubmit,
-            receiver:
-              orderToSubmit.receiver === ethers.constants.AddressZero
-                ? undefined
-                : orderToSubmit.receiver,
-          },
-          owner
-        );
+      if (deleteConditionalOrder) {
+        deletedOrders.push(conditionalOrder);
```
Comment (anxolin) [inline actions/watch.ts:122]: I moved the deletion out of the loop. I would think is not a good idea in general to mutate the collection you are iterating while you are interesting on it

File: actions/watch.ts [hunk 2]
```diff
@@ -83,101 +97,123 @@ export const checkForAndPlaceOrder: ActionFn = async (
...
 
-          await placeOrder(
-            { ...orderToSubmit, from: owner, signature },
-            chainContext.api_url
-          );
+  // Update the registry
+  await registry.write();
+};
+
+async function _checkForAndPlaceOrder(
```
Comment (anxolin) [inline actions/watch.ts:135]: This was not necesary, but I felt the main loop was doing way too many things, so I broke down the logic a bit. It was hard for me to reason about some errors I was getting (see PR description)

File: actions/watch.ts [hunk 3]
```diff
@@ -83,101 +97,123 @@ export const checkForAndPlaceOrder: ActionFn = async (
...
-                `Single order on safe ${owner} not authed. Unfilled orders:`
-              );
-            case "ProofNotAuthed":
-              console.log(
-                `Proof on safe ${owner} not authed. Unfilled orders:`
-              );
-          }
-          printUnfilledOrders(conditionalOrder.orders);
-          console.log("Removing conditional order from registry");
-          conditionalOrders.delete(conditionalOrder);
```
Comment (anxolin) [inline actions/watch.ts:171]: I don't do the deletion here anymore, i just signal to the main loop this conditional order should be removed, so:
-  we don't mutate the `conditionalOrders` (we are iterating over them)
- Its super easy to miss this logic when reading. I just made it explicit so is part of this method signature to tell the main loop when it should delete an order

File: actions/watch.ts [hunk 4]
```diff
@@ -83,101 +97,123 @@ export const checkForAndPlaceOrder: ActionFn = async (
...
+    console.error(`Unexpected error while processing order: ${e?.message}`);
   }
 
-  // Update the registry
-  await registry.write();
-};
+  return { deleteConditionalOrder: false };
+}
+
+function getOrderUid(network: string, orderToSubmit: Order, owner: string) {
```
Comment (anxolin) [inline actions/watch.ts:199]: reducing complexity of main method

File: actions/watch.ts [hunk 5]
```diff
@@ -220,32 +256,37 @@ async function placeOrder(order: any, api_url: string) {
...
+      );
+      console.log(`[placeOrder] API response`, { status, data });
     }
   } catch (error: any) {
+    const errorMessag = "[placeOrder] Error placing order in API";
     if (error.response) {
+      const { status, data } = error.response;
       // The request was made and the server responded with a status code
       // that falls out of the range of 2xx
-      console.log(JSON.stringify(error.response));
```
Comment (anxolin) [inline actions/watch.ts:238]: this was creating a cyclic dependency error

File: tenderly.yaml [hunk 6]
```diff
@@ -1,6 +1,9 @@
-account_id: 365d1c08-1187-453c-ace5-c0cc4e53a41f
+account_id: 4edfc3a8-de04-463e-8408-34302bfa01e0
+project_slug: dev@cow.fi
+provider: ""
+
 actions:
-  mfw78/rndlabs:
+  devcow/project:
```
Comment (anxolin) [inline tenderly.yaml:6]: our own Tenderly

File: tenderly.yaml [hunk 7]
```diff
@@ -1,6 +1,9 @@
-account_id: 365d1c08-1187-453c-ace5-c0cc4e53a41f
+account_id: 4edfc3a8-de04-463e-8408-34302bfa01e0
+project_slug: dev@cow.fi
```
Comment (anxolin) [inline tenderly.yaml:2]: our own Tenderly

File: actions/watch.ts (hunk 1 above)
Comment (mfw78) [reply actions/watch.ts:122]: Great catch, makes sense!

File: actions/watch.ts [hunk 8]
```diff
@@ -56,21 +57,34 @@ export const checkForSettlement: ActionFn = async (
...
-
-  console.log(
-    `Updated registry: ${JSON.stringify(
-      Array.from(registry.ownerOrders.entries())
-    )}`
-  );
   await registry.write();
 };
 
+async function getTradeableOrderWithSignature(
```
Comment (mfw78) [inline actions/watch.ts:70]: Nice!

File: actions/watch.ts [hunk 9]
```diff
@@ -83,101 +97,123 @@ export const checkForAndPlaceOrder: ActionFn = async (
...
   const chainContext = await ChainContext.create(context, blockEvent.network);
-
-  console.log(`Processing block ${blockEvent.blockNumber}...`);
+  const { network } = blockEvent;
 
   // enumerate all the owners
   for (const [owner, conditionalOrders] of registry.ownerOrders.entries()) {
-    console.log(`Checking ${owner}...`);
-
+    const deletedOrders = [];
```
Comment (mfw78) [inline actions/watch.ts:104]: nit: perhaps use `ordersPendingDelete` or similar

File: actions/watch.ts (hunk 3 above)
Comment (mfw78) [reply actions/watch.ts:171]: Yep, this makes it significantly more readable / understandable. Good work.

File: actions/watch.ts [hunk 10]
```diff
@@ -197,9 +233,9 @@ export const printUnfilledOrders = (orders: Map<BytesLike, OrderStatus>) => {
 /**
  * Place a new order
  * @param order to be placed on the cow protocol api
- * @param api_url rest api url
+ * @param apiUrl rest api url
  */
-async function placeOrder(order: any, api_url: string) {
+async function placeOrder(order: any, apiUrl: string) {
```
Comment (mfw78) [inline actions/watch.ts:238]: The :crab: within is real :laughing:

File: actions/watch.ts (hunk 5 above)
Comment (mfw78) [reply actions/watch.ts:238]: :no_bicycles:

**Review by mfw78 — APPROVED** [review]
Looks great. I'll submit a PR on the env vars to isolate those that are for the forge scripts.

**Review by mfw78 — APPROVED** [review]

File: .env.example [hunk 11]
```diff
@@ -1,7 +1,12 @@
-ETHERSCAN_API_KEY=
-PRIVATE_KEY=
-SETTLEMENT=0x9008D19f58AAbD9eD0D60971565AA8510560ab41
+# CLI
 ETH_RPC_URL=http://erigon.dappnode:8545
-SAFE=
-TWAP=
-COMPOSABLE_COW=
```
[the remark this answers is not in the document]
Comment (mfw78) [reply .env.example:7]: See commit https://github.com/cowprotocol/composable-cow/pull/35/commits/33c3ca8658b54b44d5572c3facbde701d33c827b for forge script vars.

[the remark this answers is not in the document]
Comment (anxolin) [reply .env.example:7]: Thanks!

**Review by mfw78 — APPROVED** [review]

File: actions/watch.ts [hunk 12]
```diff
@@ -220,32 +257,41 @@ async function placeOrder(order: any, api_url: string) {
...
+          headers: {
+            "Content-Type": "application/json",
+            accept: "application/json",
+          },
+        }
+      );
+      console.log(`[placeOrder] API response`, { status, data });
     }
   } catch (error: any) {
+    const errorMessag = "[placeOrder] Error placing order in API";
```
Comment (mfw78) [inline actions/watch.ts:279]: nit: `errorMessage` not `errorMessag`.

**Review by mfw78 — APPROVED** [review]
Looks great, just the minor typo nit.

**Pull request merged**